src/OrchardCore/OrchardCore.Mvc.Core/LocationExpander/ComponentViewLocationExpanderProvider.cs

```diff
@@ -56,18 +58,40 @@ public class ComponentViewLocationExpanderProvider : IViewLocationExpanderProvid
                         {
                             modulesWithComponentViews.Add(module);
                         }
+                    }
+
+                    _modulesWithComponentViews = modulesWithComponentViews;
+                }
+            }
 
-                        moduleComponentsViewFilePaths = fileProviderAccessor.FileProvider.GetViewFilePaths(
+            if (_modulesWithPagesComponentViews != null)
+            {
+                return;
```

Comment:
So, because at the beginning line 36 we still have the following, we may return if `_modulesWithComponentViews` is initialized while `_modulesWithComponentViews` may not be initalized yet.

            if (_modulesWithComponentViews != null)
            {
                return;
            }

So the suggestion is to use another `private bool _initialized` field to do the double check around the lock, then initialize both collections in the same lock and where you can re-use the same `orderedModules` computed once, then set `_initialized` to true at the end after having initialized both static collections.

Reply:
I was thinking about that, but I kept everything as its to avoid undesired results